real_time/add-custom-dynamic-entity-data-source/src/main/java/com/esri/samples/add_custom_dynamic_entity_data_source/SimulatedDataSource.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2023 Esri.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.esri.samples.add_custom_dynamic_entity_data_source;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.gson.Gson;
+
+import com.esri.arcgisruntime.data.Field;
+import com.esri.arcgisruntime.geometry.Point;
+import com.esri.arcgisruntime.geometry.SpatialReferences;
+import com.esri.arcgisruntime.realtime.DynamicEntityDataSource;
+import com.esri.arcgisruntime.realtime.DynamicEntityDataSourceInfo;
+
+/**
+ * A custom DynamicEntityDataSource for processing observations read from a given json file.
+ */
+class SimulatedDataSource extends DynamicEntityDataSource {
+  private final String fileName;
+  private final String entityIdFieldName;
+  private final long delay;
+  private final ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+  private ScheduledFuture<?> observationProcessing;
+  private final AtomicInteger currentLineNumber = new AtomicInteger();
+  private final List<String> allJsonLines = new ArrayList<>();
+  private final Gson gson = new Gson();
+
+  /**
+   * Construct a custom DynamicEntityDataSource.
+   *
+   * @param fileName name of a file with json observation data
+   * @param entityIdFieldName the name of the field containing values that uniquely identifies each entity
+   * @param delay millisecond delay between observations being added to the data source
+   */
+  SimulatedDataSource(String fileName, String entityIdFieldName, long delay) {
+    this.fileName = fileName;
+    this.entityIdFieldName = entityIdFieldName;
+    this.delay = delay;
+  }
+
+  @Override
+  protected CompletableFuture<Void> onConnectAsync() {
+    // Store all lines from file in a list
+    try (BufferedReader bufferedReader = new BufferedReader(new FileReader(fileName))) {
+      String line;
+      while ((line = bufferedReader.readLine()) != null) {
+        allJsonLines.add(line);
+      }
```

Comment:
Nothing wrong with what we have at the moment; would something like this be preferred?
```suggestion
      allJsonLines.addAll(bufferedReader.lines().filter(Objects::nonNull).toList());
```

Reply:
Much neater - I've changed to use your suggestion.